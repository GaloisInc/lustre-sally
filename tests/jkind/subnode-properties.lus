-- This file has one top level property which requires subnode
-- properties in order to prove. It's somewhat obfuscated to avoid
-- automatic invariant generation.

node counter(rand : bool) returns (x : int);
var
  lemma : bool;
let
  x = (if rand then -1 else -1) -> 1 + pre x;

  lemma = x >= -1;
  --%PROPERTY lemma;
tel;

node counter2(rand : bool) returns (x : int);
var
  lemma : bool;
let
  x = (if rand then -1 else -1) -> 1 + pre x;

  lemma = x >= -1;
  --%PROPERTY lemma;
tel;

node ten(rand : bool) returns (ten : bool);
let
  ten = counter(rand) >= 10;
tel;

node main(rand : bool) returns (prop : bool);
let
  prop = counter2(rand) < 10 or ten(rand);
  --%PROPERTY prop;
tel;
