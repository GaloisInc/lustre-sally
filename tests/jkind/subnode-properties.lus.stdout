[Lustre] Loading model from: "subnode-properties.lus"
Sally Model
===========

(define-state-type
   S
   (
     (counter Int)
     (counter2 Int)
     (lemma Bool)
     (prop Bool)
     (rand Bool)
     (ten Bool)
     (|counter2:init| Bool)
     (|counter:init| Bool)
     (|gal-initializing| Bool)
     (|lemma:1| Bool)
     (|pre_counter2| Int)
     (|pre_counter| Int)
     (|rand:1| Bool)
     (|rand:2| Bool)
     (|rand:3| Bool))
   (
     (|rand:3| Bool)))
(define-transition-system
   TS
   S
   (and
      (= |gal-initializing| true)
      (= |counter2:init| true)
      (= |counter:init| true))
   (and
      (= next.|gal-initializing| false)
      (= next.|rand:3| input.|rand:3|)
      (= next.|rand:2| next.|rand:3|)
      (= next.|pre_counter2| state.counter2)
      (ite
         state.|counter2:init|
         (=
            next.counter2
            (ite
               next.|rand:2|
               (- 1)
               (- 1)))
         (=
            next.counter2
            (+ 1 next.|pre_counter2|)))
      (= next.|counter2:init| false)
      (= next.|rand:1| next.|rand:3|)
      (= next.rand next.|rand:1|)
      (= next.|pre_counter| state.counter)
      (ite
         state.|counter:init|
         (=
            next.counter
            (ite
               next.rand
               (- 1)
               (- 1)))
         (=
            next.counter
            (+ 1 next.|pre_counter|)))
      (= next.|counter:init| false)
      (=
         next.ten
         (<= 10 next.counter))
      (=
         next.prop
         (or
            (< next.counter2 10)
            next.ten))
      (=
         next.|lemma:1|
         (<=
            (- 1)
            next.counter2))
      (=
         next.lemma
         (<=
            (- 1)
            next.counter))))

(query
   TS
   (or |gal-initializing| prop))
(query
   TS
   (or |gal-initializing| lemma))
(query
   TS
   (or |gal-initializing| |lemma:1|))

[Lustre] Validating properties:
[Lustre] Property prop...
  [Sally][Unknown]
  [Sally][Unknown]
[Lustre] Property lemma...
  [Sally][Valid] using 1 past state
[Lustre] Property lemma...
  [Sally][Valid] using 1 past state
[Lustre] Summary:
  [Valid] 2
  [Unknown] 1
  [Invalid] 0
[Lustre] Model status: [Unknown]
